### ISSUE TITLE
JSDoc optional/required variant doesn't play well with multi-@param RecordType declarations

### ISSUE THREAD

**Issue by Madd0g** [Microsoft/TypeScript#23397]
**TypeScript Version:**  2.9.0-dev.20180330

**Search Terms:**  as mentioned in [this comment](https://github.com/Microsoft/TypeScript/issues/19645#issuecomment-377952907), I'm opening this issue because with this JSDoc syntax typescript doesn't recognize the params correctly.

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
// Test this by running `tsc` on the command-line, rather than through another build tool such as Gulp, Webpack, etc.

/**
 * @param {!object} args 
 * @param {!string} args.prop1
 * @param {?string} args.prop2
 */
function patate(args) {
  const { prop1, prop2 } = args;
} 
```

**Expected behavior:** No errors

**Actual behavior:** With allowJs and checkJs, I get JSDoc '@param' tag has name 'prop1', but is no parameter with that name.

**Playground Link:**  

**Related Issues:** https://github.com/Microsoft/TypeScript/issues/19645

**Comment by RyanCavanaugh** [comment]
This just issues a different confusing error now
```
a.js:6:21 - error TS8032: Qualified name 'args.prop1' is not allowed without a leading '@param {object} args'.

6  * @param {!string} args.prop1
                      ~~~~~~~~~~

a.js:7:21 - error TS8032: Qualified name 'args.prop2' is not allowed without a leading '@param {object} args'.

7  * @param {?string} args.prop2
                      ~~~~~~~~~~
```
Changing `!object` to `object` makes everything happy but I don't think you should have to?
